ethpm/tools/builder.py

```diff
@@ -0,0 +1,332 @@
+
+    If `package_root_dir` is not provided, cwd is expected to resolve the relative
+    path to the source as defined in the compiler output.
+    """
+    return _inline_source(name, compiler_output, package_root_dir)
+
+
+@curry
+def _inline_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    package_root_dir: str,
+    manifest: Manifest,
+) -> Manifest:
+    names_and_paths = get_names_and_paths(compiler_output)
+    cwd = Path(os.getcwd())
+    source_path_suffix = names_and_paths[name]
+
+    if (cwd / source_path_suffix).is_file():
+        source_data = (cwd / source_path_suffix).read_text()
+    elif package_root_dir and (package_root_dir / source_path_suffix).is_file():
+        source_data = (package_root_dir / source_path_suffix).read_text()
+    else:
+        raise ManifestBuildingError(
+            "Contract source cannot be resolved, please make sure that the working "
+            "directory is set to the correct directory or provide `pacakge_root_dir`."
+        )
+
+    if "sources" not in manifest:
+        manifest["sources"] = {}
+
+    return assoc(
+        manifest,
+        "sources",
+        assoc(manifest["sources"], "./{0}".format(source_path_suffix), source_data),
+    )
+
+
+def pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str = None,
+) -> Manifest:
+    """
+    Pins source to IPFS and returns a copy of manifest with added field to
+    "sources": {relative_source_path: IFPS URI}.
+
+    If `package_root_dir` is not provided, cwd is expected to resolve the relative path
+    to the source as defined in the compiler output.
+    """
+    return _pin_source(name, compiler_output, ipfs_backend, package_root_dir)
+
+
+@curry
+def _pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str,
+    manifest: Manifest,
+) -> Manifest:
+    names_and_paths = get_names_and_paths(compiler_output)
+    source_path = names_and_paths[name]
+    if package_root_dir:
+        (ipfs_data,) = ipfs_backend.pin_assets(package_root_dir / source_path)
+    else:
+        cwd = Path(os.getcwd())
+        (ipfs_data,) = ipfs_backend.pin_assets(cwd / source_path)
+
+    if "sources" not in manifest:
+        manifest["sources"] = {}
+
+    return assoc(
+        manifest,
+        "sources",
+        assoc(
+            manifest["sources"],
+            "./{0}".format(source_path),
+            "ipfs://{0}".format(ipfs_data["Hash"]),
+        ),
+    )
+
+
+#
+# Contract Types
+#
+
+
+def contract_type(name: str, compiler_output: Dict[str, Any], **kwargs) -> Manifest:
+    """
+    Returns a copy of manifest with added contract_data field as specified by kwargs.
+    If no kwargs are present, all available contract_data found in the compiler output
+    will be included.
+
+    To include specific contract_data fields, add kwarg set to True (i.e. `abi=True`)
```

Comment:
Maybe instead of using `kwargs` for this they should just be explicitly listed?  This is under the assumption that we can enumerate the possible desired keys.